tests/test_ipmitool.py

```diff
@@ -0,0 +1,213 @@
+        {"name": "Settable Thresholds", "value": ""},
+        {"name": "Threshold Read Mask", "value": "lcr lnc"},
+        {"name": "Assertion Events", "value": ""},
+        {"name": "Assertions Enabled", "value": "lnc- lcr-"},
+        {"name": "Deassertions Enabled", "value": "lnc- lcr-"},
+    ],
+}]
+
+ipmitool_lan_print = """
+Set in Progress         : Set Complete
+Auth Type Support       : MD5
+Auth Type Enable        : Callback : MD5
+                        : User     : MD5
+                        : Operator : MD5
+                        : Admin    : MD5
+                        : OEM      :
+IP Address Source       : Static Address
+IP Address              : 172.16.1.2
+Subnet Mask             : 255.255.254.0
+MAC Address             : f8:bc:12:12:13:14
+SNMP Community String   : public
+IP Header               : TTL=0x40 Flags=0x40 Precedence=0x00 TOS=0x10
+BMC ARP Control         : ARP Responses Enabled, Gratuitous ARP Disabled
+Gratituous ARP Intrvl   : 2.0 seconds
+Default Gateway IP      : 172.16.210.1
+Default Gateway MAC     : 00:00:00:00:00:00
+Backup Gateway IP       : 0.0.0.0
+Backup Gateway MAC      : 00:00:00:00:00:00
+802.1q VLAN ID          : Disabled
+802.1q VLAN Priority    : 0
+RMCP+ Cipher Suites     : 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14
+Cipher Suite Priv Max   : Xaaaaaaaaaaaaaa
+                        :     X=Cipher Suite Unused
+                        :     c=CALLBACK
+                        :     u=USER
+                        :     o=OPERATOR
+                        :     a=ADMIN
+                        :     O=OEM
+Bad Password Threshold  : Not Available
+"""
+
+ipmitool_lan_print_expected = [
+    {"name": "IP Address Source", "value": "Static Address"},
+    {"name": "IP Address", "value": "172.16.1.2"},
+    {"name": "Subnet Mask", "value": "255.255.254.0"},
+    {"name": "MAC Address", "value": "f8:bc:12:12:13:14"},
+    {
+        "name": "BMC ARP Control",
+        "value": "ARP Responses Enabled, Gratuitous ARP Disabled",
+    },
+    {"name": "Default Gateway IP", "value": "172.16.210.1"},
+    {"name": "802.1q VLAN ID", "value": "Disabled"},
+    {"name": "802.1q VLAN Priority", "value": "0"},
+    {"name": "RMCP+ Cipher Suites", "value": "0,1,2,3,4,5,6,7,8,9,10,11,12,13,14"},
+]
+
+SENSOR_DATA_CMD = ["ipmitool", "sdr", "list"]
+IPMITOOL_NORMAL_ERROR = (
+    "Could not open device at /dev/ipmi0 or /dev/ipmi/0 or /dev/ipmidev/0: "
+    "No such file or directory"
+)
+IPMITOOL_ABNORMAL_ERROR = "Unexpected Error!!!"
+
+
+@mock.patch("ipmitool.run_command", autospec=True)
+class TestIpmitool:
+    def test_sensor_normal_error(self, run_command):
+        run_command.side_effect = ProcessException(
+            1,
+            SENSOR_DATA_CMD,
+            "",
+            IPMITOOL_NORMAL_ERROR,
+        )
+
+        try:
+            _ = sensor_data(None, None)
+        except XenAPIPlugin.Failure as e:
+            assert e.params == ["1", "IPMI not available", "", ""]
```

Comment:
```suggestion
        with pytest.raises(XenAPIPlugin.Failure) as e:
            _ = sensor_data(None, None)
            assert e.params == ["1", "IPMI not available", "", ""]
```

Reply:
I want to test is that XenAPIPlugin.Failure is raised by `sensor_data`  when the `run_command` raises a process exception. It looks like that with your suggestion it is the pytest that raises the XenAPIPlugin.Failure. Or maybe pytest.raises() only catches such exception?

Reply:
I think I misunderstood the usage of `with pytest.raises()`. In fact it is like an assert that the function in the block will raise such exception. I will try that it should work :+1: